File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        
        d = {}
        
        def dfs(n):
            if not n:
                return 
            d[n] = []
            
            for x in n.neighbors:
                if x not in d:
                    dfs(x)
                    
                d[n].append(x)
                
        def dfscreate(n,d):
            if not n:
                return Node()
            
            node = Node(n.val)
            vis[n] = node
            for x in d[n]:
                if x not in vis:
                    node.neighbors.append(dfscreate(x,d))
                else:
                    node.neighbors.append(vis[x])
            
            return node
            
        dfs(node)
        vis = {}
        
        return dfscreate(node,d) if d else None
        

File: 0133-clone-graph/0133-clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        unordered_map<Node*, Node*> mp;
        return dfs(node,mp);
    }
    
    Node* dfs(Node* root, unordered_map<Node*, Node*> &mp){
        if(!root)
            return NULL;
        
        Node* node = new Node(root->val);
        mp[root] = node;
        
        for(auto n:root->neighbors){
            if(mp.find(n)==mp.end())
                node->neighbors.push_back(dfs(n,mp));
            else
                node->neighbors.push_back(mp[n]);
        }
        
        return node;
        
    }
};